src/optimizer/optimizer_tasks.py

```diff
@@ -0,0 +1,254 @@
+    def task_type(self):
+        return self._task_type
+
+    @property
+    def optimizer_context(self):
+        return self._optimizer_context
+
+    @root_expr.setter
+    def root_expr(self, expr):
+        self.root_expr = expr
+
+    def execute(self):
+        raise NotImplementedError
+
+
+class TopDownRewrite(OptimizerTask):
+    def __init__(self, root_expr: GroupExpression,
+                 optimizer_context: OptimizerContext):
+        super().__init__(root_expr, root_expr.group_id,
+                         optimizer_context, OptimizerTaskType.TOP_DOWN_REWRITE)
+
+    @property
+    def root_expr(self):
+        return self._root_expr
+
+    @root_expr.setter
+    def root_expr(self, expr):
+        self._root_expr = expr
+
+    def execute(self):
+        """We apply rewrite rules in a top down fashion.
+        Right now we are applying rules aggressively. Later
+        when we have more rules it might be a better idea to
+        push optimization task to a queue.
+        """
+        rewrite_rules = RulesManager().rewrite_rules
+        valid_rules = []
+        for rule in rewrite_rules:
+            if not self.root_expr.is_rule_explored(rule.rule_type) and rule.top_match(self.root_expr.opr):
+                valid_rules.append(rule)
+
+        # sort the rules by promise
+        valid_rules = sorted(valid_rules, key=lambda x: x.promise(), reverse=True)
+        for rule in valid_rules:
+            binder = Binder(self.root_expr, rule.pattern,
+                            self.optimizer_context.memo)
+            for match in iter(binder):
+                if not rule.check(match, self.optimizer_context):
+                    continue
+                # Uncertain correctness
+                self.root_expr.mark_rule_explored(rule.rule_type)
+                LoggingManager().log('In TopDown, Rule {} matched for {}'
+                                     .format(rule, self.root_expr),
+                                     LoggingLevel.INFO)
+                after = rule.apply(match, self.optimizer_context)
+                new_expr = self.optimizer_context.xform_opr_to_group_expr(
+                    opr=after,
+                    root_group_id=self.root_expr.group_id,
+                    is_root=True,
+                    copy_opr=False
+                )
+                self.root_expr = new_expr
+                LoggingManager().log('After rewiting {}'
+                                     .format(self.root_expr),
+                                     LoggingLevel.INFO)
+                self.optimizer_context.task_stack.push(TopDownRewrite(
+                    self.root_expr, self.optimizer_context))
+
+        for child in self.root_expr.children:
+            child_expr = self.optimizer_context.memo.groups[child] \
+                .logical_exprs[0]
+            self.optimizer_context.task_stack.push(TopDownRewrite(
+                child_expr, self.optimizer_context))
+
+
+class BottomUpRewrite(OptimizerTask):
```

Comment:
@xzdandy @gaurav274 
What is the necessity of BottomUpRewrite here? given we already have a top down rewrite optimization.

Reply:
We need multiple iterations to apply all rules. But it is a good question what is the difference between applying merely top down rewrite multiple times and applying top down and bottom up alternatively.

Reply:
I see. Because those papers that I skimmed through only picked one way to apply rules, I am just curious why we need both directions here.

Reply:
Consider the logical tree as A->B->C->D. In top-down, we start from A and keep applying rules as we go down. So, if let's say we merge B and C to C', we never go back to A and check if certain rules that weren't valid earlier(A->B) are valid now(A->C', since the subtree changed). So the idea is to quickly apply the majority of the rewrite rules. 
But in the bottom-up, we start from D and keep going up. So, if we rewrite let say B->C to C', we again try to apply rules to (C'->D). So, it is expensive but thorough.

Reply:
@jiashenC The optimizer discussed in those papers is the part we run after doing the top-down and bottom-up rewrite. Right now, we aren't doing much there. All the cost-based interesting stuff happens there. That part at a high level is just top-down.

Reply:
I see. That makes sense.